url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/MOVEM.java

```diff
@@ -0,0 +1,692 @@
+				}
+			}
+		}
+	}
+
+	protected final int movem_word_r2m(int opcode)
+	{
+		int reglist = cpu.fetchPCWord();
+		Operand dst = cpu.resolveDstEA((opcode >> 3) & 0x07, opcode & 0x07, Size.Word);
+		int address = dst.getComputedAddress();
+		int count;
+
+		//true up if -(An) addressing mode
+		if(((opcode >> 3) & 0x07) == 4)
+		{
+			count = putMultipleWordPreDec(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = putMultipleWord(reglist, address);
+		}
+
+		return R2M_Timing[dst.index()] + (count << 2);	//reg count * 4
+	}
+
+	protected final int movem_long_r2m(int opcode)
+	{
+		int reglist = cpu.fetchPCWord();
+		Operand dst = cpu.resolveDstEA((opcode >> 3) & 0x07, opcode & 0x07, Size.Long);
+		int address = dst.getComputedAddress();
+		int count;
+
+		//true up if -(An) addressing mode
+		if(((opcode >> 3) & 0x07) == 4)
+		{
+			count = putMultipleLongPreDec(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = putMultipleLong(reglist, address);
+		}
+
+		return R2M_Timing[dst.index()] + (count << 3);	//reg count * 8
+	}
+
+	protected final int movem_word_m2r(int opcode)
+	{
+		int reglist = cpu.fetchPCWord();
+		Operand src = cpu.resolveSrcEA((opcode >> 3) & 0x07, opcode & 0x07, Size.Word);
+		int address = src.getComputedAddress();
+		int count;
+
+		// true up if (An)+ addressing mode
+		if(((opcode >> 3) & 0x07) == 3)
+		{
+			count = getMultipleWordPostInc(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = getMultipleWord(reglist, address);
+		}
+
+		return M2R_Timing[src.index()] + (count << 2);	//reg count * 4
+	}
+
+	protected final int movem_long_m2r(int opcode)
+	{
+		int reglist = cpu.fetchPCWord();
+		Operand src = cpu.resolveSrcEA((opcode >> 3) & 0x07, opcode & 0x07, Size.Long);
+		int address = src.getComputedAddress();
+		int count;
+
+		//true up if (An)+ addressing mode
+		if(((opcode >> 3) & 0x07) == 3)
+		{
+			count = getMultipleLongPostInc(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = getMultipleLong(reglist, address);
+		}
+
+		return M2R_Timing[src.index()] + (count << 3);	//reg count * 8
+	}
+
+	public final DisassembledInstruction disassembleOp(int address, int opcode, Size sz, boolean reg_to_mem)
+	{
+		DisassembledOperand src;
+		DisassembledOperand dst;
+		int mode = (opcode >> 3) & 0x07;
+		int reg = (opcode & 0x07);
+		int reglist = cpu.readMemoryWord(address + 2);
+		boolean reversed = (mode == 4);
+
+		if(reg_to_mem)
+		{
+			//registers to memory
+			src = new DisassembledOperand(regListToString(reglist, reversed), 2, reglist);
+			dst = cpu.disassembleDstEA(address + 4, mode, reg, sz);
+		}
+		else
+		{
+			// memory to registers
+			src = cpu.disassembleSrcEA(address + 4, mode, reg, sz);
+			dst = new DisassembledOperand(regListToString(reglist, reversed), 2, reglist);
+		}
+
+		return new DisassembledInstruction(address, opcode, "movem" + sz.ext(), src, dst);
+	}
+
+	protected final String regListToString(int reglist, boolean reversed)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.